cmd/operator/internal/controller/kubearchiveconfig_controller.go

```diff
@@ -46,24 +50,51 @@ type KubeArchiveConfigReconciler struct {
 //+kubebuilder:rbac:groups=core,resources=serviceaccounts,verbs=create;delete;get;list;update;watch
 //+kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=roles;rolebindings,verbs=bind;create;delete;escalate;get;list;update;watch
 //+kubebuilder:rbac:groups=sources.knative.dev,resources=apiserversources,verbs=create;delete;get;list;update;watch
+//+kubebuilder:rbac:groups="",resources=configmaps,verbs=create;delete;get;list;update;watch
 
 func (r *KubeArchiveConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	log := log.FromContext(ctx)
 
 	log.Info("KubeArchiveConfig reconciling.")
 
 	kaconfig := &kubearchivev1alpha1.KubeArchiveConfig{}
-	err := r.Get(ctx, req.NamespacedName, kaconfig)
-	if err != nil {
-		if errors.IsNotFound(err) {
-			log.Info("KubeArchiveConfig resource not found. Ignoring since object must have been deleted.")
-			return ctrl.Result{}, nil
+	if err := r.Get(ctx, req.NamespacedName, kaconfig); err != nil {
+		// Ignore not-found errors, since they can't be fixed by an immediate requeue (we need
+		// to wait for a new notification), and we can get them on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	finalizerName := "kubearchive.org/finalizer"
+
+	if kaconfig.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, add the finalizer if necessary.
+		if !controllerutil.ContainsFinalizer(kaconfig, finalizerName) {
+			controllerutil.AddFinalizer(kaconfig, finalizerName)
+			if err := r.Update(ctx, kaconfig); err != nil {
+				return ctrl.Result{}, err
+			}
 		}
-		log.Error(err, "Failed to get KubeArchiveConfig, requeuing the request.")
-		return ctrl.Result{}, err
+	} else {
+		// The object is being deleted.
+		if controllerutil.ContainsFinalizer(kaconfig, finalizerName) {
+			// Finalizer is present, clean up filters from ConfigMap.
+			if err := r.removeFilters(ctx, kaconfig); err != nil {
+				// If filter deletion fails, return with error so that it can be retried.
+				return ctrl.Result{}, err
+			}
+
+			// Remove the finalizer from the list and update it.
+			controllerutil.RemoveFinalizer(kaconfig, finalizerName)
+			if err := r.Update(ctx, kaconfig); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+
+		// Stop reconciliation as the resource is being deleted.
+		return ctrl.Result{}, nil
 	}
 
-	_, err = r.reconcileA14eServiceAccount(ctx, kaconfig)
+	_, err := r.reconcileA14eServiceAccount(ctx, kaconfig)
```

Comment:
Why do you need a colon here? `err` is defined on line 61 of the new code.

Reply:
The compiler gives an error. I believe the `err` defined on line 61 is scoped to the `if` statement and not outside.

Reply:
Ah, okay. Thanks for checking.